design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,345 @@
+was available before translation to determine what the error represents and to decide what to do with it.
+With the Try Pattern, custom domain specific errors can be passed back as parameters if necessary.
+The parameter type and its possible values are declared in the producing component and would be projected
+across languages as is. This means the consumer can rely on using it to make decisions without being concerned
+of the translation accuracy.
+
+But, this isn't to say that all errors should be passed back via the Try Pattern. If a component intends
+on returning / throwing an error which a consumer may use to control code flow, then the Try Pattern should be
+used. For errors which are considered fatal like out of memory and invalid argument, it would be fine to
+return / throw the error as that would allow for useful diagnostic information to be collected about it which
+would be available upon failure.
+
+## Diagnosing application failures
+
+Errors coming from components written in other languages can be one of the causes of application
+failures. Just like with any other failure, developers will want to diagnose where and why the
+failure occurred. To assist with this, xlang will provide the ability for a projection to associate
+diagnostic information with the error. The following information can be associated:
+
+- Language specific error (i.e. error code, exception type)
+- Error message
+- Stack trace
+- Projection identifier
+- Any additional diagnostic information
+
+Some of this information can be represented by different languages in different ways. For instance,
+in some languages, stack traces can be an array of addresses while in others it can be a list of
+function names. Due to this, xlang will store the information generically as strings and it will
+be the responsibility of the projections to convert the representation it has to a string representation
+that can be logged and is readable by a developer or tooling.
+
+Errors may also end up traveling through several different components written in different languages
+before becoming a failure. To assist with diagnosing these failures, xlang will allow a projection to
+associate additional diagnostic information to the existing information as an error travels through it.
+This will help a developer get a complete picture of the failure which wouldn't be available with just
+the information collected from the language that initially returned the error. An example of this is
+the stack trace. A language will typically only be aware of the stack within its language boundary and
+thereby the diagnostic information collected at the initial error would only have a partial stack.
+The following additional information can be associated with the error as it travels through projections:
+
+- Stack trace
+- Projection identifier
+- Any additional diagnostic information
+
+The diagnostic information will be stored in a COM object (error info) implementing the
```

Comment:
Should we try to avoid mentioning COM except when comparing systems? This would be an XLangObject really right?

Reply:
as per arch meeting, this should be using IXlangObject

Reply:
This is now a IXlangObject.